Android app crashes upon push notification

Android app installed from published release, i.e. built with Release configuration, crashes when it receives a notification.
Launching the app itself works fine.

Logs from the generated bug report:

```
08-07 01:10:37.619 1000 3821 3875 I ActivityManager: Start proc 23516:dev.zoechbauer.nudelsieb/u0a382 for broadcast dev.zoechbauer.nudelsieb/com.google.firebase.iid.FirebaseInstanceIdReceiver
08-07 01:10:37.634 10382 23516 23516 E bauer.nudelsie: Not starting debugger since process cannot load the jdwp agent.
08-07 01:10:37.636 1000 3821 3988 I AutomaticBrightnessControllerEx: [MultiALC] lux = 44.0, time = 361608124
08-07 01:10:37.674 10382 23516 23516 I bauer.nudelsie: The ClassLoaderContext is a special shared library.
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: Unable to instantiate appComponentFactory
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: java.lang.ClassNotFoundException: Didn't find class "androidx.core.app.CoreComponentFactory" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk"],nativeLibraryDirectories=[/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/lib/arm64, /data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64, /system/vendor/lib64]]
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.app.LoadedApk.createAppFactory(LoadedApk.java:232)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.app.LoadedApk.createOrUpdateClassLoaderLocked(LoadedApk.java:737)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.app.LoadedApk.getClassLoader(LoadedApk.java:816)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.app.LoadedApk.getResources(LoadedApk.java:1038)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.app.ContextImpl.createAppContext(ContextImpl.java:2385)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6038)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.app.ActivityThread.access$1100(ActivityThread.java:208)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1696)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.os.Handler.dispatchMessage(Handler.java:106)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.os.Looper.loop(Looper.java:205)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at android.app.ActivityThread.main(ActivityThread.java:6991)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at java.lang.reflect.Method.invoke(Native Method)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
08-07 01:10:37.687 10382 23516 23516 E LoadedApk: at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:884)
08-07 01:10:37.804 10382 23516 23516 W monodroid-gc: GREF GC Threshold: 46080
08-07 01:10:37.929 1000 4668 4668 D UsbHostManagerService: boot not completed yet.
08-07 01:10:37.948 10382 23516 23516 I FirebaseApp: Device unlocked: initializing all Firebase APIs for app [DEFAULT]
08-07 01:10:37.963 10382 23516 23516 I FirebaseInitProvider: FirebaseApp initialization successful
08-07 01:10:37.970 10382 23516 23516 D AndroidRuntime: Shutting down VM
FATAL
EXCEPTION
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: FATAL EXCEPTION: main
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: Process: dev.zoechbauer.nudelsieb, PID: 23516
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: java.lang.RuntimeException: Unable to instantiate receiver com.google.firebase.iid.FirebaseInstanceIdReceiver: java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.iid.FirebaseInstanceIdReceiver" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk"],nativeLibraryDirectories=[/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/lib/arm64, /data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64, /system/vendor/lib64]]
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at android.app.ActivityThread.handleReceiver(ActivityThread.java:3485)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at android.app.ActivityThread.access$1200(ActivityThread.java:208)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1707)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at android.os.Handler.dispatchMessage(Handler.java:106)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at android.os.Looper.loop(Looper.java:205)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at android.app.ActivityThread.main(ActivityThread.java:6991)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at java.lang.reflect.Method.invoke(Native Method)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:884)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.iid.FirebaseInstanceIdReceiver" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk"],nativeLibraryDirectories=[/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/lib/arm64, /data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64, /system/vendor/lib64]]
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at android.app.AppComponentFactory.instantiateReceiver(AppComponentFactory.java:84)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: at android.app.ActivityThread.handleReceiver(ActivityThread.java:3478)
08-07 01:10:37.971 10382 23516 23516 E AndroidRuntime: ... 8 more
08-07 01:10:37.974 10106 5477 5477 W GCM : broadcast intent callback: result=CANCELLED forIntent { act=com.google.android.c2dm.intent.RECEIVE pkg=dev.zoechbauer.nudelsieb (has extras) }
08-07 01:10:37.988 1000 3821 10604 D DisplayManagerService: Display listener for pid 21572 died.
08-07 01:10:38.043 10382 23516 23516 I MonoDroid: UNHANDLED EXCEPTION:
08-07 01:10:38.072 10382 23516 23516 I MonoDroid: Java.Lang.RuntimeException: Unable to instantiate receiver com.google.firebase.iid.FirebaseInstanceIdReceiver: java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.iid.FirebaseInstanceIdReceiver" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk"],nativeLibraryDirectories=[/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/lib/arm64, /data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64, /system/vendor/lib64]] ---> Java.Lang.ClassNotFoundException: Didn't find class "com.google.firebase.iid.FirebaseInstanceIdReceiver" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk"],nativeLibraryDirectories=[/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/lib/arm64, /data/app/dev.zoechbauer.nu
08-07 01:10:38.072 10382 23516 23516 I MonoDroid: --- End of inner exception stack trace ---
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: --- End of managed Java.Lang.RuntimeException stack trace ---
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: java.lang.RuntimeException: Unable to instantiate receiver com.google.firebase.iid.FirebaseInstanceIdReceiver: java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.iid.FirebaseInstanceIdReceiver" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk"],nativeLibraryDirectories=[/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/lib/arm64, /data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64, /system/vendor/lib64]]
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at android.app.ActivityThread.handleReceiver(ActivityThread.java:3485)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at android.app.ActivityThread.access$1200(ActivityThread.java:208)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1707)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at android.os.Handler.dispatchMessage(Handler.java:106)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at android.os.Looper.loop(Looper.java:205)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at android.app.ActivityThread.main(ActivityThread.java:6991)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at java.lang.reflect.Method.invoke(Native Method)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:884)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.firebase.iid.FirebaseInstanceIdReceiver" on path: DexPathList[[zip file "/system/framework/org.apache.http.legacy.boot.jar", zip file "/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk"],nativeLibraryDirectories=[/data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/lib/arm64, /data/app/dev.zoechbauer.nudelsieb-89s5Bj8MpgOwzRXhoi4-gQ==/base.apk!/lib/arm64-v8a, /system/lib64, /system/product/lib64, /system/vendor/lib64]]
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:134)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at android.app.AppComponentFactory.instantiateReceiver(AppComponentFactory.java:84)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: at android.app.ActivityThread.handleReceiver(ActivityThread.java:3478)
08-07 01:10:38.073 10382 23516 23516 I MonoDroid: ... 8 more
08-07 01:10:38.073 10382 23516 23516 I MonoDroid:
08-07 01:10:38.082 1000 3821 3873 I ActivityManager: Showing crash dialog for package dev.zoechbauer.nudelsieb u0
```

--- a/src/Nudelsieb/Nudelsieb.Mobile/Nudelsieb.Mobile.Android/FirebaseService.cs
+++ b/src/Nudelsieb/Nudelsieb.Mobile/Nudelsieb.Mobile.Android/FirebaseService.cs
@@ -52,8 +52,6 @@ namespace Nudelsieb.Mobile.Droid
         /// </summary>
         public override void OnNewToken(string token)
         {
-            // TODO: save token instance locally, or log if desired
-
             SendRegistrationToServer(token);
         }
 

--- a/src/Nudelsieb/Nudelsieb.Mobile/Nudelsieb.Mobile.Android/Nudelsieb.Mobile.Android.csproj
+++ b/src/Nudelsieb/Nudelsieb.Mobile/Nudelsieb.Mobile.Android/Nudelsieb.Mobile.Android.csproj
@@ -25,6 +25,10 @@
     <AndroidSupportedAbis />
     <AppPackageNamePostFix>.debug</AppPackageNamePostFix>
     <AppDisplayName>1 Nudelsieb Debug</AppDisplayName>
+    <AotAssemblies>false</AotAssemblies>
+    <EnableLLVM>false</EnableLLVM>
+    <AndroidEnableProfiledAot>false</AndroidEnableProfiledAot>
+    <BundleAssemblies>false</BundleAssemblies>
   </PropertyGroup>
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
     <DebugSymbols>false</DebugSymbols>
@@ -38,8 +42,17 @@
     <AndroidLinkTool>r8</AndroidLinkTool>
     <EmbedAssembliesIntoApk>true</EmbedAssembliesIntoApk>
     <AndroidKeyStore>false</AndroidKeyStore>
-    <AppPackageNamePostFix></AppPackageNamePostFix>
+    <AppPackageNamePostFix>
+    </AppPackageNamePostFix>
     <AppDisplayName>Nudelsieb</AppDisplayName>
+    <AotAssemblies>false</AotAssemblies>
+    <EnableLLVM>false</EnableLLVM>
+    <AndroidEnableProfiledAot>false</AndroidEnableProfiledAot>
+    <BundleAssemblies>false</BundleAssemblies>
+    <AndroidLinkMode>SdkOnly</AndroidLinkMode>
+    <MandroidI18n />
+    <AndroidLinkSkip>
+    </AndroidLinkSkip>
   </PropertyGroup>
   <PropertyGroup>
     <AndroidApplication>True</AndroidApplication>
@@ -105,6 +118,7 @@
   </ItemGroup>
   <ItemGroup>
     <GoogleServicesJson Include="$(GoogleServicesJsonFilePath)" />
+    <ProguardConfiguration Include="proguard.cfg" />
     <None Include="Resources\AboutResources.txt" />
     <None Include="Assets\AboutAssets.txt" />
     <None Include="Properties\AndroidManifest.xml" />

--- a/src/Nudelsieb/Nudelsieb.Mobile/Nudelsieb.Mobile.Android/proguard.cfg
+++ b/src/Nudelsieb/Nudelsieb.Mobile/Nudelsieb.Mobile.Android/proguard.cfg
@@ -1,0 +1,3 @@
+﻿-dontwarn com.google.android.gms.**
+-keep class com.google.android.gms.** { *; }
+-keep class com.google.firebase.** { *; }

--- a/src/Nudelsieb/Nudelsieb.Mobile/Nudelsieb.Mobile/AppSettings.cs
+++ b/src/Nudelsieb/Nudelsieb.Mobile/Nudelsieb.Mobile/AppSettings.cs
@@ -50,7 +50,7 @@ namespace Nudelsieb.Mobile
             {
                 Console.WriteLine(
                     $"Error, could not apply {nameof(AppSettings)} from {SecretFileName} " +
-                    $"({ex.Message})");
+                    $"({ex.GetType()}: {ex.Message})");
             }
 
             return appSettings;

--- a/src/Nudelsieb/Nudelsieb.Mobile/Nudelsieb.Mobile/Directory.Build.targets
+++ b/src/Nudelsieb/Nudelsieb.Mobile/Nudelsieb.Mobile/Directory.Build.targets
@@ -2,7 +2,8 @@
 <Project>
     <Target Name="AddUserSecrets"
             BeforeTargets="PrepareForBuild"
-            Condition=" '$(Configuration)' == 'Debug' And '$(UserSecretsId)' != '' ">
+            Condition=" '$(UserSecretsId)' != '' ">
+        <!-- also do this for Release configuration, to verify correct linking of ProGuard/r8 -->
         <PropertyGroup>
             <UserSecretsFilePath Condition=" '$(OS)' == 'Windows_NT' ">
                 $([System.Environment]::GetFolderPath(SpecialFolder.UserProfile))\AppData\Roaming\Microsoft\UserSecrets\$(UserSecretsId)\secrets.json

--- a/src/Nudelsieb/Nudelsieb.Shared.Clients/Authentication/AuthenticationService.cs
+++ b/src/Nudelsieb/Nudelsieb.Shared.Clients/Authentication/AuthenticationService.cs
@@ -125,6 +125,7 @@ namespace Nudelsieb.Shared.Clients.Authentication
                     return account;
                 }
             }
+
             return null!; // todo non null
         }
     }

